# 데이터 로딩 및 가공

In [6]:
import pandas as pd
import numpy as np
import warnings; warnings.filterwarnings('ignore')

movies = pd.read_csv('/content/tmdb_5000_movies.csv')
print(movies.shape)
movies.head(1)

(4803, 20)


budget  \
0  237000000   

                                                                                                genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                                                                              keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "sp...   

  original_language original_title  \
0                en         Avatar   

                                                                                              overview  \
0  In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, ...   

   popularity  \
0  150.437577   

                                                                                  production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289}, {"name": "Twentieth Century Fox Film Corporatio...   

                                                                                  production_countries  \
0  [{"iso_3166_1": "US", "name": "United States of America"}, {"iso_3166_1": "GB", "name": "United ...   

  release_date     revenue  runtime  \
0   2009-12-10  2787965087    162.0   

                                                                        spoken_languages  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso_639_1": "es", "name": "Espa\u00f1ol"}]   

     status                      tagline   title  vote_average  vote_count  
0  Released  Enter the World of Pandora.  Avatar           7.2       11800

- 4803개의 레코드, 20개의 피처
- 다양한 메타 정보 중 콘텐츠 기반 필터링 추천 분석에 사용할 주요 칼럼만 추출
  - id, title, genres, vote_average, vote_count, popularity, keywords, overview

In [7]:
movies_df = movies[['id', 'title', 'genres', 'vote_average', 'vote_count', 'popularity', 'keywords', 'overview']]

- genres, keywords 등의 칼럼: 리스트 내붸 여러 개의 딕셔너리가 있는 형태의 문자열로 표기 => 가공 필요

In [8]:
pd.set_option('max_colwidth', 100)
movies_df[['genres', 'keywords']][:1]

genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {...   

                                                                                              keywords  
0  [{"id": 1463, "name": "culture clash"}, {"id": 2964, "name": "future"}, {"id": 3386, "name": "sp...

- 여러 개의 개별 데이터를 가짐, 딕셔너리의 키로 추출 가능
- 문자열을 분해해서 개별 데이터를 파이썬 리스트 객체로 추출
- ast 모듈의 literal_eval() 함수로 문자열을 list[dict1, dict2] 객체로 만들 수 있음

In [9]:
from ast import literal_eval
movies_df['genres'] = movies_df['genres'].apply(literal_eval)
movies_df['keywords'] = movies_df['keywords'].apply(literal_eval)

- name 키에 해당하는 값만 리스트 객체로 추출

In [10]:
movies_df['genres'] = movies_df['genres'].apply(lambda x : [y['name'] for y in x])
movies_df['keywords'] = movies_df['keywords'].apply(lambda x : [y['name'] for y in x])
movies_df[['genres', 'keywords']][:1]

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                                                                              keywords  
0  [culture clash, future, space war, space colony, society, space travel, futuristic, romance, spa...

# 장르 콘텐츠 유사도 측정

- genres 칼럼을 count 기반 피처 벡터화 변환
- 변환한 데이터 세트를 코사인 유사도를 통해 비교
- 장르 유사도가 높은 영화 중 평점이 높은 순으로 추천

In [14]:
from sklearn.feature_extraction.text import CountVectorizer

movies_df['genres_literal'] = movies_df['genres'].apply(lambda x:(' ').join(x))
count_vect = CountVectorizer(min_df=0, ngram_range=(1,2))
genre_mat = count_vect.fit_transform(movies_df['genres_literal'])
print(genre_mat.shape)

(4803, 276)


- 4803개의 레코드, 276개의 개별 단어 피처로 구성된 피처 벡터 행렬 생성
- 코사인 유사도 계산
  - cosine_similarity(): 기준 행과 비교 행의 코사인 유사도를 행렬 형태로 반환

In [15]:
from sklearn.metrics.pairwise import cosine_similarity

genre_sim = cosine_similarity(genre_mat, genre_mat)
print(genre_sim.shape)
print(genre_sim[:1])

(4803, 4803)
[[1.         0.59628479 0.4472136  ... 0.         0.         0.        ]]


- genre_sim의 기준 행별로 비교 대상이 되는 행의 유사도 값이 높은 순으로 정렬된 행렬의 위치 인덱스 값 추출

In [17]:
genre_sim_sorted_ind = genre_sim.argsort()[:, ::-1]
print(genre_sim_sorted_ind[:1])

[[   0 3494  813 ... 3038 3037 2401]]


- 자신(0) 제외, 3494번 레코드가 가장 유사도가 높고, 그 다음 813, 가장 낮은 레코드는 2401

# 장르 콘텐츠 필터링을 이용한 영화 추천

- 장르 유사도에 따라 영화 추천하는 함수

In [1]:
def find_sim_movie(df, sorted_ind, title_name, top_n=10):
  title_movie = df[df['title'] == title_name]

  title_index = title_movie.index.values
  similar_indexes = sorted_ind[title_index, :(top_n)]

  print(similar_indexes)
  similar_indexes = similar_indexes.reshape(-1)

  return df.iloc[similar_indexes]

In [19]:
similar_movies = find_sim_movie(movies_df, genre_sim_sorted_ind, 'The Godfather', 10)
similar_movies[['title', 'vote_average']]

[[2731 1243 3636 1946 2640 4065 1847 4217  883 3866]]


title  vote_average
2731                          The Godfather: Part II           8.3
1243                                    Mean Streets           7.2
3636                                   Light Sleeper           5.7
1946  The Bad Lieutenant: Port of Call - New Orleans           6.0
2640         Things to Do in Denver When You're Dead           6.7
4065                                      Mi America           0.0
1847                                      GoodFellas           8.2
4217                                            Kids           6.8
883                              Catch Me If You Can           7.7
3866                                     City of God           8.1

- '대부'와 유사한 영화 추천 결과
  - 비슷한 유형인 '대부 2편', '좋은 친구' 추천
  - '라이트 슬리퍼', 'Mi America' 등의 영화 추천은 개선 필요
- 더 많은 후보군 선정 후 평점에 따라 필터링해서 최종 추천하는 방식으로 변경
  - 'vote_average': 소수 관객이 부여한 평점에 왜곡된 데이터

In [20]:
movies_df[['title', 'vote_average', 'vote_count']].sort_values('vote_average', ascending=False)[:10]

title  vote_average  vote_count
3519          Stiff Upper Lips          10.0           1
4247     Me You and Five Bucks          10.0           2
4045     Dancer, Texas Pop. 81          10.0           1
4662            Little Big Top          10.0           1
3992                 Sardaarji           9.5           2
2386            One Man's Hero           9.3           2
2970        There Goes My Baby           8.5           2
1881  The Shawshank Redemption           8.5        8205
2796     The Prisoner of Zenda           8.4          11
3337             The Godfather           8.4        5893

- 가중평점 방식을 사용
  - (v/(v+m)) * R + (m/(v+m)) * C

In [21]:
C = movies_df['vote_average'].mean()
m = movies_df['vote_count'].quantile(0.6)
print('C: ', round(C,3), 'm: ', round(m,3))

C:  6.092 m:  370.2


- 기존 평점을 새로운 가중 평점으로 변경하는 함수 생성
- 새 평점 정보 'vote_weighted' 생성

In [24]:
percentile = 0.6
m = movies['vote_count'].quantile(percentile)
C = movies['vote_average'].mean()

def weighted_vote_average(record):
  v = record['vote_count']
  R = record['vote_average']

  return ((v/(v+m)) * R + (m/(v+m)) * C)

movies_df['weighted_vote'] = movies.apply(weighted_vote_average, axis = 1)

In [26]:
movies_df[['title','vote_average','weighted_vote','vote_count']].sort_values('weighted_vote', ascending=False)[:10]

title  vote_average  weighted_vote  vote_count
1881  The Shawshank Redemption           8.5       8.396052        8205
3337             The Godfather           8.4       8.263591        5893
662                 Fight Club           8.3       8.216455        9413
3232              Pulp Fiction           8.3       8.207102        8428
65             The Dark Knight           8.2       8.136930       12002
1818          Schindler's List           8.3       8.126069        4329
3865                  Whiplash           8.3       8.123248        4254
809               Forrest Gump           8.2       8.105954        7927
2294             Spirited Away           8.3       8.105867        3840
2731    The Godfather: Part II           8.3       8.079586        3338

- 장르 유사성이 높은 영화를 top_n의 2배수만큼 후보군으로 선정한 후 weighted_vote 칼럼 값이 높은 순으로 추출

In [33]:
def find_sim_movie(df, sorted_ind, title_name, top_n=10):
  title_movie = df[df['title'] == title_name]
  title_index = title_movie.index.values

  similar_indexes = sorted_ind[title_index, :(top_n*2)]
  similar_indexes = similar_indexes.reshape(-1)

  similar_indexes = similar_indexes[similar_indexes != title_index]

  return df.iloc[similar_indexes].sort_values('weighted_vote', ascending=False)[:top_n]

similar_movies = find_sim_movie(movies_df, genre_sim_sorted_ind, 'The Godfather', 10)
similar_movies[['title', 'vote_average', 'weighted_vote']]

title  vote_average  weighted_vote
2731       The Godfather: Part II           8.3       8.079586
1847                   GoodFellas           8.2       7.976937
3866                  City of God           8.1       7.759693
1663  Once Upon a Time in America           8.2       7.657811
883           Catch Me If You Can           7.7       7.557097
281             American Gangster           7.4       7.141396
4041              This Is England           7.4       6.739664
1149              American Hustle           6.8       6.717525
1243                 Mean Streets           7.2       6.626569
2839                     Rounders           6.9       6.530427